### ISSUE TITLE
How to generate MAC?

### ISSUE THREAD

**Issue by ganduras** [DansMaCulotte/monetico-php#1]
In your code MAC is a static string but is required to generate dynamic string

**Comment by Zevran** [comment]
Hi,

MAC field is a generated string from payment information and security key.
Could you detail your usage ?

**Comment by ganduras** [comment]
I want to use monetico as payment method. But before I want to understand how its working. For this reason Im asking

**Comment by ganduras** [comment]
Yes I found that is using method hash_hmac, so it's dynamic

**Comment by Zevran** [comment]
MAC is generated by [generateSeal](https://github.com/DansMaCulotte/monetico-php/blob/master/src/Payment/Payment.php#L151) method.
To understant of Monetico works you should read their [documentation](https://www.monetico-paiement.fr/fr/info/documentations/Monetico_Paiement_documentation_technique_v1_0j.pdf).

**Comment by ganduras** [comment]
I already did it but it have many mistakes. Another question: in which case you have to use getUsableKey() ?
Thanks

**Comment by Zevran** [comment]
You don't have to use directly the `generateSeal` method.
You don't have to use getUsableKey either, but to answer to your question it is needed every time it is used MAC generation and MAC validation.

**Comment by ganduras** [comment]
What is the purpose of getUsableKey ? I tried to understand the code but its stange , many substrings...
The function is receiving merchant security key and after ?

**Comment by Zevran** [comment]
It comes from Monetico, they want your security formated that way to generate seal and to validate seal. I didn't produce this code, it comes from their code example.

**Comment by ganduras** [comment]
What about checking the payment status? the curl url is the same? How you found values of : bincb , vld... ?

**Comment by LupusVII** [comment]
Hi !
Monetico send everything to your CGI URL in POST.
So to get the vars, you just have to write $_POST['bincb'].

First, you must regenerate the monetico object (like the exemple given) :
```php
use DansMaCulotte\Monetico\Monetico;

$monetico = new Monetico(
    'EPT_CODE',
    'SECURITY_KEY',
    'COMPANY_CODE',
    'RETURN_URL',
    'RETURN_SUCCESS_URL',
    'RETURN_ERROR_URL'
);
```

Then, you build the data array, like this :
```php
$data = [
                    'date' => $_POST['date'],
                    'amount' => $_POST['montant'],
                    'reference' => $_POST['reference'],
                    'MAC' => $_POST['MAC'],
                    'texte-libre' => $_POST['texte-]ibre'),
                    'code-retour' => $_POST['code-]etour'),
                    'cvx' => $_POST['cvx'],
                    'vld' => $_POST['vld'],
                    'brand' => $_POST['brand'],
                    'status3ds' => $_POST['status3ds'],
                    'numauto' => $_POST['numauto'],
                    'originecb' => $_POST['originecb'],
                    'bincb' => $_POST['bincb'],
                    'hpancb' => $_POST['hpancb'],
                    'ipclient' => $_POST['ipclient'],
                    'originetr' => $_POST['originetr'],
                    'veres' => $_POST['veres'] ?: "",
                    'pares' => $_POST['pares'] ?: ""
                ];
```
You create a Response object :
```php
$response = new \DansMaCulotte\Monetico\Payment\Response($data);
```
You validate the seal :
```php
$result = $monetico->validateSeal($response);
```

You generate the expected receipt :
```php
$receipt = new \DansMaCulotte\Monetico\Payment\Receipt($result);
```

If you want to do specific stuff, now you just have to test the response code :
```php
if ('payetest' == $response->returnCode || 'paiement' == $response->returnCode) {
  // do stuff if payment is ok
}
else {
  // do other stuff if payment went wrong
}
```

Print the receipt value just by doing this :
```php
echo $receipt;
die;
```

And you are done :D

**Comment by ganduras** [comment]
LupusVII, do you know something about debug mode? I listened that you need to send them the link where you will work, so it's impossible to send if you are working in localhost

**Comment by LupusVII** [comment]
Totally. It's not the best system to work with. A lot of things must be sent directly to the bank. 
Your work MUST BE online so their server can call it. If you are on localhost, they cannot reach it and they cannot deploy your account into production mode.

Put your website online for the payment tests and once your payment system is okay, ask for production deployment (make sure you made 3 test payments with correct receipts as the doc ask). Wait for the approval and once you got it, put the website offline and keep working on localhost.
Everything will be ready for the "real launch" ;)

**Comment by Zevran** [comment]
Thanks for answering @LupusVII ! @ganduras You can use [ngrok](https://ngrok.com/) or [localtunnel](https://localtunnel.github.io/www/) to make it work locally. I use ngrok !

**Comment by lovecoding-git** [comment]
Hi @LupusVII,

In your code, MAC is received from post?
I've submit the form but Monetico says the MAC is not correct.
Please explain in detail.

**Comment by LupusVII** [comment]
Hi @lovecoding-git 
There is not a lot more to explain. Monetico says send you a POST request on the URL set up (last time I did it, I had to tell them the URL by email, there is no field to set up it in there backend).
I think you do not have an issue with this step but maybe with the previous one? When you generate the payment form?

Also, it might be best to create a new issue than reopening an old one :)
